### PR TITLE
chore: extract and refactor typings schematypeoptions

### PR DESCRIPTION
**Pull request by Uzlopak**
Regarding #11482

### CODE REVIEW

**Comment by Uzlopak** [comment]
Dont review/merge yet. I am not yet very satisfied with the changes.

File: test/types/schema.test.ts [hunk 1]
```diff
@@ -1,10 +1,10 @@
-import { Schema, Document, SchemaDefinition, Model } from 'mongoose';
+import { Schema, Document, SchemaDefinition, Model, SchemaTypes } from 'mongoose';
 import { expectType, expectNotType, expectError } from 'tsd';
 
 enum Genre {
-  Action,
-  Adventure,
-  Comedy
+  Action = 'Action',
```
Comment (vkarpov15) [inline test/types/schema.test.ts:5]: Is this change necessary? I kinda like the default `enum` definition. Not a big deal either way since this is test code.

File: types/schematypeoptions.d.ts [hunk 2]
```diff
@@ -0,0 +1,219 @@
...
+  interface ValidateOpts<T> {
+    msg?: string;
+    message?: string | ValidatorMessageFn<T>;
+    type?: string;
+    validator: ValidateFn<T> | LegacyAsyncValidateFn<T> | AsyncValidateFn<T>;
+  }
+
+  type SchemaValidator<T> = RegExp | [RegExp, string] | Function | [Function, string] | ValidateOpts<T> | ValidateOpts<T>[];
+
+  export class SchemaTypeOptionsClass implements
```
Comment (vkarpov15) [inline types/schematypeoptions.d.ts:33]: This seems a little strange. Why not just use `typeof SchemaTypeOptions` like we currently do?

File: types/schematypeoptions.d.ts [hunk 3]
```diff
@@ -0,0 +1,219 @@
...
+    StringSchemaTypeOptions,
+    NumberSchemaTypeOptions,
+    ObjectIdSchemaTypeOptions,
+    DateSchemaTypeOptions,
+    MapSchemaTypeOptions,
+    BufferSchemaTypeOptions,
+    BasicSchemaTypeOptions<any> {
+  }
+
+  export type SchemaTypeOptions<T> =
```
Comment (vkarpov15) [inline types/schematypeoptions.d.ts:43]: Why the separate type? It seems like `BasicSchemaTypeOptions` does similar `T extends string` logic.
